pkg/core/workflow/types.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package workflow includes all type definitaions about Workflow.
+package workflow
+
+import (
+	"github.com/jinzhu/gorm"
+)
+
+// Workflow defines the root structure of a workflow.
+type Workflow struct {
+	gorm.Model
+	UID         string         `gorm:"index:uid" json:"uid"`
+	Name        string         `json:"name"`
+	Entry       string         `json:"entry"` // the entry node name
+	Status      workflowStatus `json:"status"`
+	CurrentNode TopologyNode   `json:"current_node"`
+	Topology    Topology       `json:"topology"`
+}
+
+// workflowState defines the current state of a workflow.
+//
+// Includes: Initializing, Running, Errored, Finished.
+//
+// Const definitions can be found below this type.
+type workflowStatus string
+
+const (
+	// WorkflowInitializing represents a workflow is being initialized.
+	WorkflowInitializing workflowStatus = "Initializing"
+
+	// WorkflowRunning represents that a workflow is running.
+	WorkflowRunning workflowStatus = "Running"
+
+	// WorkflowErrored represents an error in a workflow.
+	WorkflowErrored workflowStatus = "Errored"
+
+	// WorkflowFinished represents that a workflow has ended.
+	WorkflowFinished workflowStatus = "Finished"
+)
+
+// Topology describes the process of a workflow.
+type Topology struct {
+	Nodes []TopologyNode `json:"nodes"`
+}
+
+// TopologyNode defines the basic structure of a node.
+type TopologyNode struct {
+	Type  nodeType  `json:"type"`
+	State nodeState `json:"state,omitempty"`
+}
+
+// Node defines the single step of a workflow.
+type Node struct {
+	TopologyNode
+	Serial   NodeSerial   `json:"serial,omitempty"`
+	Parallel NodeParallel `json:"parallel,omitempty"`
+	Template Template     `json:"template"`
+}
+
+// NodeSerial defines SerialNode's specific fields.
+type NodeSerial struct {
+	Tasks    []Node `json:"tasks"`
```

Comment:
I prefer `Tasks []string`.

Because we already put all the Node in the field `Workflow.Topology` as an array. If we using a struct here, there might a little redundancy in our JSON.

Reply:
The `Workflow.Topology` is just a skeleton of a Node. It only includes `Type` and `State` (can be `nil` when creating a workflow). I think I forgot to define an API about getting the complete nodes of a workflow:

- `GET` `/workflows/{id}/nodes`

If we defined it as `[]string`, we can only get the `task` detail from the template (by its name string). According to my current understanding, a `Node` is an instance of a `Template`. One design for runtime, the other is a static template.

Reply:
`Node` also has a unique name. There might some `Node`s that refs to one same `Template`, we use node name to distinguish.

Does the sorting of items in `Workflow.Topology` is required? If YES, what's the expected order?

Reply:
> There might some Nodes that refs to one same Template, we use node name to distinguish.

🙋🏽‍♂️ I think I understand. I will change `[]Node` to `[]string` later.

> Does the sorting of items in Workflow.Topology is required? If YES, what's the expected order?

Yes. Should be sorted from entry to the end.

Reply:
Done.

Reply:
> Yes. Should be sorted from entry to the end.

But the topology is a tree, we could not describe it only by this array. 🤔

Reply:
> 🙋🏽‍♂️ I think I understand. I will change []Node to []string later.

I think we should a new field `name` in `TopologyNode`. After that, both `[]string` and `[]Node` are OK to me.

Reply:
> I think we should a new field name in TopologyNode.

Already added in 7e1f08a.

Reply:
After discussing with @STRRL. This conversation's content is outdated.

We removed the `TopologyNode` and put its fields into `Node`.

Please refer to the latest code. I will mark this conversation as resolved.